sources/openeuler-http_test.go

```diff
@@ -0,0 +1,62 @@
+package sources
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetLatestRelease(t *testing.T) {
+	s := &openEuler{}
+
+	tests := []struct {
+		release string
+		want    string
+	}{
+		{
+			"22.03",
+			"22.03-LTS",
+		},
+		{
+			"20.03",
+			"20.03-LTS-SP3",
+		},
+		{
+			"20.03-LTS",
+			"20.03-LTS-SP3",
+		},
+		{
+			"20.03-LTS-SP1",
+			"20.03-LTS-SP1",
+		},
+		{
+			"21.03",
+			"21.03",
+		},
+	}
+
+	for _, test := range tests {
+		release, err := s.getLatestRelease("https://repo.openeuler.org/", test.release)
+		require.NoError(t, err)
+		require.NotEmpty(t, release)
+		require.Equal(t, test.want, release)
+	}
+}
+
+func TestGetLatestReleaseFailed(t *testing.T) {
+	s := &openEuler{}
+
+	tests := []string{
+		"22.00",      // non-existed release
+		"BadRelease", // invalid format
+		"",           // null string
+	}
+
+	for _, release := range tests {
+		_, err := s.getLatestRelease("https://repo.openeuler.org/", release)
+		require.NotNil(t, err)
+	}
+
+	_, err := s.getLatestRelease("foobar", "20.03")
+	require.NotNil(t, err)
+}
```

Comment:
Instead of adding a separate test function, you can just add a new variable `shouldFail` (or something like that) to the other struct. We do this in [sources/common_test.go](https://github.com/lxc/distrobuilder/blob/master/sources/common_test.go#L32).

Reply:
Reworked :)